### PR TITLE
feat: add, edit, delete cards

### LINKED ISSUE #75: [FEATURE]: add, edit, delete cards
## Summary

- Implementation of add, edit and delete card features

## User story it relates to

#72 

### Motivation

We need this to allow the creation, updating and deletion of cards by the user

### PR DESCRIPTION
**Pull request by nunocaseiro**
Relates to #75

## Proposed Changes

  - HOC to protect the routes that require autentication
  - Refactoring the code in order to use the board's new model 
  - Add card feature
  - Edit card feature
  - Remove card feature
  - Replace yup by zod
  - Alert dialog for confirming the delete action
  - Transformboard helper to modify optimistically the board when a card is moved and dropped
  - Added the request to add, edit, delete the cards in the backend
  - Added the request to update the card position in the backend
  - Refactor nextauth to save the user's id
  - Package.json updated
  
This pull request is intendend to close #75

### CODE REVIEW

File: backend/src/models/boards/boards.service.ts [hunk 1]
```diff
@@ -231,7 +230,10 @@ export default class BoardsService {
           },
         },
         {
-          arrayFilters: [{ 'c._id': cardId }, { 'item._id': cardItemId }],
+          arrayFilters: [
+            { 'c._id': cardId },
```
Comment (RubenMCCarreira) [inline backend/src/models/boards/boards.service.ts:234]: what is `c._id` ? can you be more explicit pls?

File: frontend/types/next-auth.d.ts [hunk 2]
```diff
@@ -7,32 +7,33 @@ declare module "next-auth" {
     user: {
       name: string;
       email: string;
+      id: string;
```
Comment (RubenMCCarreira) [inline frontend/types/next-auth.d.ts:10]: in this new type fies, I saw `_id` and now I see just `id`. these types have to be different, or you could keep the consistency?

File: backend/src/models/boards/boards.service.ts (hunk 1 above)
Comment (nunocaseiro) [reply backend/src/models/boards/boards.service.ts:234]: image

As we can see in the above picture, I am updating the text of a card [line 228] but I need to filter the array to find the card that I want. So I used  a letter "c" to identify a element of the card's array. What I am doing in array filters is saying that in the array of cards of that column I want the card "c" with that Id.

File: frontend/utils/fetchData.tsx [hunk 3]
```diff
@@ -8,13 +9,29 @@ export const instance = axios.create({
   },
 });
 
+const nonNeededToken = ["/auth/login", "/auth/refresh"];
+
+instance.interceptors.request.use(async (config) => {
+  const { url, headers } = config;
+  if (url && headers) {
+    if (!nonNeededToken.includes(url)) {
```
Comment (RubenMCCarreira) [inline frontend/utils/fetchData.tsx:17]: can you joint the two statements ?

File: frontend/types/next-auth.d.ts (hunk 2 above)
Comment (nunocaseiro) [reply frontend/types/next-auth.d.ts:10]: The other file is in backend. This is frontend. I really don't like `_id` but in mongo is the default :)

File: frontend/utils/fetchData.tsx (hunk 3 above)
Comment (nunocaseiro) [reply frontend/utils/fetchData.tsx:17]: yes, ofc

File: backend/src/models/boards/boards.service.ts (hunk 1 above)
Comment (RubenMCCarreira) [reply backend/src/models/boards/boards.service.ts:234]: oh ok ok, I understood. but using just `c`, to a new guy like me in the project, I have no ideia what is it xD

Comment (nunocaseiro) [reply backend/src/models/boards/boards.service.ts:234]: Sure, I will rename it to `card` instead of a letter xd

**Review by RubenMCCarreira — DISMISSED** [review]

File: frontend/types/next-auth.d.ts (hunk 2 above)
Comment (RubenMCCarreira) [reply frontend/types/next-auth.d.ts:10]: oh ok, sry xD

File: backend/src/models/boards/boards.service.ts (hunk 1 above)
Comment (RubenMCCarreira) [reply backend/src/models/boards/boards.service.ts:234]: ty

**Review by RubenMCCarreira — APPROVED** [review]

**Pull request merged**